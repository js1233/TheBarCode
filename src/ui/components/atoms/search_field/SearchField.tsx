import Cross from "assets/images/cross.svg";
import Search from "assets/images/search.svg";
import { COLORS, FONTS, FONT_SIZE, SPACE } from "config";
import { usePreferredTheme } from "hooks";
import React, {
  MutableRefObject,
  useEffect,
  useRef,
  useState
} from "react";
import {
  StyleProp,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  View,
  ViewStyle
} from "react-native";

interface OwnProps {
  style?: StyleProp<ViewStyle>;
  placeholder: string;
  onChangeText: (textToSearch?: string) => void;
  leftIcon?: boolean;
  rightIcon?: boolean;
  borderType?: BORDER_TYPE;
  OnClickedSearchField?: () => void;
}

type Props = OwnProps;

export enum BORDER_TYPE {
  DASHED = "dashed",
  SOLID = "solid",
  DOTTED = "dotted"
}

export const SearchField = React.memo<Props>(
  ({
    placeholder,
    style,
    onChangeText,
    leftIcon = false,
    rightIcon = false,
    borderType,
    OnClickedSearchField
  }) => {
    const [currentSearchText, setCurrentSearchText] = useState("");
    const theme = usePreferredTheme();
    let isProcess: MutableRefObject<boolean> = useRef(true);

    useEffect(() => {
      const timeoutRef = setTimeout(() => {
        if (isProcess.current) {
          isProcess.current = false;
          return;
        } else {
          onChangeText(currentSearchText);
        }
      }, 1000);

      return () => {
        clearTimeout(timeoutRef);
      };
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentSearchText]);

    const getBorderStyle = () => {
      if (borderType === BORDER_TYPE.DASHED) {
        return styles.dashedBorderStyle;
      } else if (borderType === BORDER_TYPE.SOLID) {
        return styles.solidBorderStyle;
      } else if (borderType === BORDER_TYPE.DOTTED) {
        return styles.dottedBorderStyle;
      }
    };

    return (
      <View
        style={[
          getBorderStyle(),
          styles.container,
          {
            backgroundColor: theme.themedColors.primaryBackground,
            borderColor: theme.themedColors.interface["900"]
          },
          style
        ]}>
        {leftIcon && (
          <Search
            width={16}
            height={16}
            style={styles.leftIcon}
            testID={"left-icon"}
            stroke={theme.themedColors.interface["500"]}
          />
        )}
        <TextInput
          value={currentSearchText}
          placeholderTextColor={COLORS.theme?.interface["500"]}
          placeholder={placeholder}
          numberOfLines={1}
          testID="SEARCH"
          style={[
            leftIcon ? styles.textInput : styles.textInput,
            { color: COLORS.theme?.interface["900"] }
          ]}
          onTouchStart={OnClickedSearchField}
          onChangeText={setCurrentSearchText}
        />

        {rightIcon && (
          <TouchableOpacity onPress={() => setCurrentSearchText("")}>
            <Cross
              width={14}
              height={14}
              testID={"right-icon"}
              style={styles.rightIcon}
              fill={theme.themedColors.primaryColor}
            />
          </TouchableOpacity>
        )}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    height: 40,
    alignItems: "center",
    flexDirection: "row",
    justifyContent: "center",
    borderRadius: 10
  },
  leftIcon: {
    aspectRatio: 1,
    marginLeft: 12,
    marginRight: 5
  },
  rightIcon: {
    aspectRatio: 1,
    marginLeft: 8,
    marginRight: SPACE.sm
  },
  textInput: {
    fontFamily: FONTS.regular,
    flex: 1,
    fontSize: FONT_SIZE._3xs,
    paddingRight: SPACE.sm
  },
  dashedBorderStyle: {
    borderStyle: "dashed",
    borderWidth: 1
  },
  solidBorderStyle: {
    borderStyle: "solid",
    borderWidth: 1
  },
  dottedBorderStyle: {
    borderStyle: "dotted",
    borderWidth: 1
  }
});
