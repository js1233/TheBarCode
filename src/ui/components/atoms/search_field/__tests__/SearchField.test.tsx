import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { SearchField } from "../SearchField";

const searchText = () => {
  return "Muhammad noman";
};

it("renders correctly", async () => {
  const { getByTestId } = render(
    <SearchField
      placeholder="Enter text"
      onChangeText={() => searchText}
      leftIcon={false}
      rightIcon={false}
    />
  );

  const textToSearch = getByTestId("SEARCH");
  fireEvent.changeText(textToSearch, searchText);
});

it("Snapshot testing", () => {
  const rendered = render(
    <SearchField
      placeholder="Enter text"
      onChangeText={() => searchText}
      leftIcon={false}
      rightIcon={false}
    />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("Search by entered characters", async () => {
  const onChangeText = jest.fn();
  let inProcess: boolean = true;
  const { getByTestId } = render(
    <SearchField
      placeholder="Enter text"
      onChangeText={onChangeText}
      leftIcon={false}
      rightIcon={false}
    />
  );
  const textToSearch = getByTestId("SEARCH");

  await waitFor(() => {
    fireEvent.changeText(textToSearch, "Muhammad noman");
    if (inProcess) {
      inProcess = false;
      return;
    } else {
      expect(onChangeText).toBeCalledTimes(1);
      expect(onChangeText).toHaveBeenCalledWith("Muhammad noman");
    }
  });
});

it("renders left icon", () => {
  const { queryByTestId } = render(
    <SearchField
      placeholder={"Enter Text"}
      onChangeText={() => searchText}
      leftIcon={true}
      rightIcon={false}
    />
  );
  const leftIcon = queryByTestId("left-icon");
  const rightIcon = queryByTestId("right-icon");
  expect(rightIcon).toBeNull();
  expect(leftIcon).not.toBeNull();
});

it("renders right icon", () => {
  const { queryByTestId } = render(
    <SearchField
      placeholder={"Enter Text"}
      onChangeText={() => searchText}
      leftIcon={false}
      rightIcon={true}
    />
  );
  const rightIcon = queryByTestId("right-icon");
  const leftIcon = queryByTestId("left-icon");
  expect(leftIcon).toBeNull();
  expect(rightIcon).not.toBeNull();
});
