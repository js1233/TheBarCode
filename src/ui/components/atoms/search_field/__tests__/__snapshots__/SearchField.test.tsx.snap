// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "flexDirection": "row",
        "height": 40,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#171717",
      },
      undefined,
    ]
  }
>
  <TextInput
    numberOfLines={1}
    onChangeText={[Function]}
    placeholder="Enter text"
    placeholderTextColor="#737373"
    style={
      Array [
        Object {
          "flex": 1,
          "fontFamily": "ProximaNova-Regular",
          "fontSize": 22,
          "paddingRight": 8,
        },
        Object {
          "color": "#171717",
        },
      ]
    }
    testID="SEARCH"
    value=""
  />
</View>
`;
